core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts

```diff
@@ -345,6 +354,99 @@ oppia.directive('questionPlayer', [
             }
           };
 
+          ctrl.reviewAndRetry = function() {
+            if (!ctrl.failedSkillIds || ctrl.failedSkillIds.length === 0) {
+              throw Error('No failed skills');
+            }
+
+            var conceptCards = [];
+            var failedSkills = [];
+            ctrl.failedSkillIds.forEach(function(skillId) {
+              failedSkills.push(
+                ctrl.scorePerSkillMapping[skillId].description);
+            });
+            ConceptCardBackendApiService.loadConceptCards(
+              ctrl.failedSkillIds
+            ).then(function(conceptCardBackendDicts) {
+              conceptCardBackendDicts.forEach(function(conceptCardBackendDict) {
+                conceptCards.push(
+                  ConceptCardObjectFactory.createFromBackendDict(
+                    conceptCardBackendDict));
+              });
+            });
+
+            $timeout(function() {
+              $uibModal.open({
+                templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+                  '/components/concept-card/concept-card-modal.template.html'
+                ),
+                backdrop: true,
+                controller: [
+                  '$scope', '$filter', '$uibModalInstance', '$window',
+                  'UrlService',
+                  function(
+                      $scope, $filter, $uibModalInstance, $window,
+                      UrlService) {
+                    $scope.conceptCards = conceptCards;
```

Comment:
Wait I am a little confused as to why you are doing this in the question player. I thought we decided on having a separate directive for all the concept card stuff?

Reply:
I think every modal is handled like this in the codebase. I'm not sure if I'm supposed to create a separate directive for a modal.

Reply:
No you don't need a separate directive for the modal, what you can do is create a separate directive for the concept card (lets say conceptCardDirective). This directive will have the explanation and the worked examples + the javascript logic to keep showing more worked examples (i.e. everything you currently have in modal-body div in the html). In the modal html, you can do the following: 
```
<div class="modal-header">
  <h3> <[currentSkill]> </h3>
</div>
<div class="modal-body">
  <concept-card relevant_params=relevant_params></concept-card>
</div>
<div class="modal-footer">
  <button class="btn btn-default" ng-click="cancel()"> Cancel </button>
  <button class="btn btn-default" ng-if="!isLastConceptCard()" ng-click="next()"> Next </button>
  <button class="btn btn-default"
          ng-if="isLastConceptCard()"
          ng-click="retryTest()"
          style="background-color: #11776D; color: #FFFFFF;">
    Retry Test
  </button>
</div>
```
Then the <concept-card> directive can be reused everywhere

Reply:
Done.